include/deal.II/base/work_stream.h

```diff
@@ -874,14 +1019,352 @@ namespace WorkStream
+          Copier                                   copier,
+          const ScratchData &                      sample_scratch_data,
+          const CopyData &                         sample_copy_data,
+          const unsigned int                       chunk_size = 8)
+      {
+        if (MultithreadInfo::n_threads() == 1)
+          {
+            // need to copy the sample since it is marked const
+            ScratchData scratch_data = sample_scratch_data;
+            CopyData    copy_data    = sample_copy_data; // NOLINT
+
+            for (Iterator i = begin; i != end; ++i)
+              {
+                // need to check if the function is not the zero function. To
+                // check zero-ness, create a C++ function out of it and check
+                // that
+                if (static_cast<const std::function<void(
+                      const Iterator &, ScratchData &, CopyData &)> &>(worker))
+                  worker(i, scratch_data, copy_data);
+                if (static_cast<const std::function<void(const CopyData &)> &>(
+                      copier))
+                  copier(copy_data);
+              }
+
+            return;
+          }
+
+        tf::Executor &executor = MultithreadInfo::get_taskflow_executor();
+        tf::Taskflow  taskflow;
+
+        tf::Task last_copier = taskflow.placeholder();
+
+        Threads::ThreadLocalStorage<std::unique_ptr<ScratchData>>
+          thread_local_scratch;
+
+        std::vector<std::unique_ptr<Chunk<CopyData>>> chunks;
+
+        unsigned int idx             = 0;
+        unsigned int remaining_items = std::distance(begin, end);
+
+        const unsigned int real_chunk_size =
+          (remaining_items / chunk_size < 3 * MultithreadInfo::n_threads()) ?
+            1 :
+            chunk_size;
+
+        Iterator it = begin;
+        while (it != end)
+          {
+            unsigned int count  = std::min(remaining_items, real_chunk_size);
+            Iterator     middle = it;
+            std::advance(middle, count);
+
+            chunks.emplace_back();
+
+            // this chunk works on [it,middle)
+            auto worker_task =
+              taskflow
+                .emplace([it_begin = it,
+                          it_end   = middle,
+                          idx,
+                          count,
+                          &sample_scratch_data,
+                          &thread_local_scratch,
+                          &chunks,
+                          &sample_copy_data,
+                          &worker]() {
+                  auto &scratch_ptr = thread_local_scratch.get();
+                  if (!scratch_ptr.get())
+                    scratch_ptr =
+                      std::make_unique<ScratchData>(sample_scratch_data);
+
+                  ScratchData &scratch = *scratch_ptr.get();
+                  chunks[idx] =
+                    std::make_unique<Chunk<CopyData>>(count, sample_copy_data);
```

Comment:
this is the origin of the previous two questions: It's not clear to me what `idx` corresponds to, and what guarantees that `chunks` has the right size already.

Reply:
The line right here sets the size of this chunk to ``count`` items. We have N copier tasks that each operate on a chunk with ``count`` number of items. ``idx`` is the nth copier task.